pkg/net/libp2p/bootstrap.go

```diff
@@ -0,0 +1,223 @@
+
+	"github.com/jbenet/goprocess"
+	goprocessctx "github.com/jbenet/goprocess/context"
+	periodicproc "github.com/jbenet/goprocess/periodic"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/routing"
+)
+
+// ErrNotEnoughBootstrapPeers signals that we do not have enough bootstrap
+// peers to bootstrap correctly.
+var ErrNotEnoughBootstrapPeers = errors.New("not enough bootstrap peers to bootstrap")
+
+// BootstrapConfig specifies parameters used in the network bootstrapping process.
+type BootstrapConfig struct {
+	// Period governs the periodic interval at which the node will
+	// attempt to bootstrap. The bootstrap process is not very expensive, so
+	// this threshold can afford to be small (<=30s).
+	Period time.Duration
+
+	// ConnectionTimeout determines how long to wait for a bootstrap
+	// connection attempt before cancelling it.
+	ConnectionTimeout time.Duration
+
+	// BootstrapPeers is a function that returns a set of bootstrap peers
+	// for the bootstrap process to use. This makes it possible for clients
+	// to control the peers the process uses at any moment.
+	BootstrapPeers func() []peer.AddrInfo
+}
+
+// DefaultBootstrapConfig specifies default sane parameters for bootstrapping.
+var DefaultBootstrapConfig = BootstrapConfig{
+	Period:            30 * time.Second,
+	ConnectionTimeout: (30 * time.Second) / 3, // Perod / 3
+}
+
+func bootstrapConfigWithPeers(pis []peer.AddrInfo) BootstrapConfig {
+	cfg := DefaultBootstrapConfig
+	cfg.BootstrapPeers = func() []peer.AddrInfo {
+		return pis
+	}
+	return cfg
+}
+
+// Bootstrap kicks off bootstrapping. This function will periodically
+// check the number of open connections and -- if there are too few -- initiate
+// connections to well-known bootstrap peers. It also kicks off subsystem
+// bootstrapping (i.e. routing).
+func Bootstrap(
+	id peer.ID,
+	host host.Host,
+	rt routing.Routing,
+	cfg BootstrapConfig,
+) (io.Closer, error) {
+	// make a signal to wait for one bootstrap round to complete.
+	doneWithRound := make(chan struct{})
+
+	// the periodic bootstrap function -- the connection supervisor
+	periodic := func(worker goprocess.Process) {
+		ctx := goprocessctx.OnClosingContext(worker)
+
+		if err := bootstrapRound(ctx, host, cfg); err != nil {
+			logger.Warningf("bootstrap round error: [%v]", err)
+		}
+
+		<-doneWithRound
+	}
+
+	// kick off the node's periodic bootstrapping
+	proc := periodicproc.Tick(cfg.Period, periodic)
+	proc.Go(periodic) // run one right now.
+
+	// kick off Routing.Bootstrap
+	if rt != nil {
+		ctx := goprocessctx.OnClosingContext(proc)
+		if err := rt.Bootstrap(ctx); err != nil {
+			if procErr := proc.Close(); procErr != nil {
+				return nil, fmt.Errorf(
+					"process closing error [%v] while handling bootstrap error [%v]",
+					procErr,
+					err,
+				)
+			}
+			return nil, err
+		}
```

Comment:
Notice that due to `client-scan` errors I had to change this fragment from:
```
		if err := rt.Bootstrap(ctx); err != nil {
			proc.Close()
			return nil, err
		}
```
to 
```
		if err := rt.Bootstrap(ctx); err != nil {
			if procErr := proc.Close(); procErr != nil {
				return nil, fmt.Errorf(
					"process closing error [%v] while handling bootstrap error [%v]",
					procErr,
					err,
				)
			}
			return nil, err
		}
```
See https://github.com/keep-network/keep-core/pull/3051/commits/b5711bff7b7367f96e69db6a4ff07572cf5a0f55.

Also I had to change some function and variable name to lowercase to make them unexported.
See https://github.com/keep-network/keep-core/pull/3051/commits/c162488093d2c53cce8a8dde31b47c9d35dc5c3c.